modules: hal_nordic: Update nrfx to version 2.6.0

nordic-krch opened the pull request:
Update the hal_nordic module revision, to switch to nrfx v2.6.0.

nrfx 2.6.0 includes extended API in nrfx_gpiote which can be
used by the zephyr shim.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
modules: hal_nordic: Update nrfx to version 2.6.0

Update the hal_nordic module revision, to switch to nrfx v2.6.0.

nrfx 2.6.0 includes extended API in nrfx_gpiote which can be
used by the zephyr shim.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nika-nordic commented on modules/hal_nordic/nrfx/CMakeLists.txt:
```diff
@@ -58,6 +58,7 @@ zephyr_library_sources_ifdef(CONFIG_SOC_NRF5340_CPUNET ${MDK_DIR}/system_nrf5340
 zephyr_library_sources_ifdef(CONFIG_SOC_NRF9160        ${MDK_DIR}/system_nrf9160.c)
 
 zephyr_library_sources(nrfx_glue.c)
+zephyr_library_sources(${NRFX_DIR}/helpers/nrfx_flag32_allocator.c)
```
What about introducing `set(HELPERS_DIR ${NRFX_DIR}/helpers)` and using it here? There might be more helpers with .c files in the future

nordic-krch committed:
Bluetooth: Controller: nordic: Rework hw resources files

Move all defines that specifies (D)PPI resources which are used
to dedicated files which have no include dependencies. This allow
to use them in nrfx_glue.h to specify fixed (D)PPI channels used
by the bluetooth.

Additionally, added mask with GPIOTE channels used by the bluetooth
controller. It was previously not defined which could lead to
conflicts since same channels could be allocated by the nrfx_gpiote
user

Updated nrfx_glue.h to use new approach.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

anangl commented on modules/hal_nordic/nrfx/CMakeLists.txt:
```diff
@@ -58,6 +59,7 @@ zephyr_library_sources_ifdef(CONFIG_SOC_NRF5340_CPUNET ${MDK_DIR}/system_nrf5340
 zephyr_library_sources_ifdef(CONFIG_SOC_NRF9160        ${MDK_DIR}/system_nrf9160.c)
 
```
I think that for now, until the corresponding memory region is excluded from use in Zephyr and `NRF_FICR_NS` becomes actually used, `NRF_SKIP_FICR_NS_COPY_TO_RAM` should be defined in this file for `CONFIG_SOC_NRF5340_CPUAPP` and `CONFIG_SOC_NRF9160`.

nordic-krch replied:
is this related to nrfx2.6.0 or can be made in separate PR? I'm not familiar with this topic.

anangl replied:
It's a new feature that comes with the MDK included in nrfx 2.6.0 (see [here](https://github.com/zephyrproject-rtos/hal_nordic/pull/106/files#diff-5fec5f44d9e363f8a7befe0695b5a91b1f8c9ee3c2d968df350f14de8feb8931R121-R123)), so I think it should be addressed in this PR.

nordic-krch replied:
done

anangl reviewed (commented):
Why are these two commits:
- drivers: pwm: pwm_nrf5_sw: Use runtime resources allocation
- bluetooth: controller: ll_sw: nordic: Rework hw resources files

included in this PR? Are they required for updating nrfx to 2.6.0?

nordic-krch commented:
PR is based on #40615.

nrfx2.6.0 changed the way how ppi, gpiote resources masks are used. They are used to initialize static variable. Because of that `extern uint32_t flags` cannot be used because it is not considered constant by the compiler. This triggered rework how fixed channels by other modules are used in `nrfx_glue.x`:
- sw pwm dependency is removed here: #40615
- bluetooth controller: Moved all defines which are configuring ppi/gpiote channels to new headers which doesn't have other dependencies (e.g. use of nrfx hal). There are two files with ppi/dppi resources and one file aggregating both and adding gpiote configuration `radio_nrf5_resources.h`. These files can be included by `nrfx_glue.h` because they have no nrfx dependencies.
-  mpsl/radio driver added includes directly in `nrfx_glue.h`

nordic-krch commented:
>  Are they required for updating nrfx to 2.6.0?

@anangl see https://github.com/zephyrproject-rtos/zephyr/pull/40556#issuecomment-977728674

nordic-krch committed:
modules: hal_nordic: nrfx: Set NRF_SKIP_FICR_NS_COPY_TO_RAM

Set NRF_SKIP_FICR_NS_COPY_TO_RAM for nrf53 and nrf91 until corresponding
memory region is excluded from use.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

tejlmand commented on modules/hal_nordic/nrfx/CMakeLists.txt:
```diff
@@ -42,6 +42,11 @@ zephyr_compile_definitions_ifdef(CONFIG_NRF_SECURE_APPROTECT_USER_HANDLING
 zephyr_library_compile_definitions_ifdef(CONFIG_NRF_TRACE_PORT
                                  ENABLE_TRACE)
 
+zephyr_compile_definitions_ifdef(CONFIG_SOC_NRF5340_CPUAPP
```
Nothing wrong on this line, but the commit message 95616a5d029f0d3ab284bf41ab5719c82e2dfcbf says:
> memory region is ecluded from use.

I assume you mean: `excluded`

tejlmand reviewed (approved):
code looks good.

Minor spelling mistake in a commit message noticed.

nordic-krch commented:
#40615 is merged now and  is rebased.

@cvinayak can you take a look at rework in bt_ctlr headers?

anangl commented on modules/hal_nordic/nrfx/nrfx_glue.h:
```diff
@@ -237,39 +272,59 @@ void nrfx_busy_wait(uint32_t usec_to_wait);
 #endif
 
+#if NRF_802154_VERIFY_PERIPHS_ALLOC_AGAINST_MPSL
+BUILD_ASSERT(
+	(NRFX_PPI_CHANNELS_USED_BY_802154_DRV & NRFX_PPI_CHANNELS_USED_BY_MPSL) == 0,
+	"PPI channels used by the Bluetooth controller overlap with those "
+	"assigned to the pwm_nrf5_sw driver.");
```
This copy-pasted message needs to be corrected.

anangl commented on modules/hal_nordic/nrfx/nrfx_glue.h:
```diff
@@ -237,39 +272,59 @@ void nrfx_busy_wait(uint32_t usec_to_wait);
+	"PPI channels used by the Bluetooth controller overlap with those "
+	"assigned to the pwm_nrf5_sw driver.");
+
+BUILD_ASSERT(
+	(NRFX_PPI_GROUPS_USED_BY_802154_DRV & NRFX_PPI_GROUPS_USED_BY_MPSL) == 0,
+	"PPI channels used by the Bluetooth controller overlap with those "
+	"assigned to the pwm_nrf5_sw driver.");
```
Same here.

anangl commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5_ppi_resources.h:
```diff
@@ -0,0 +1,169 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef RADIO_NRF5_RESOURCES_H__
```
```suggestion
#ifndef RADIO_NRF5_PPI_RESOURCES_H__
```

anangl commented on modules/hal_nordic/nrfx/nrfx_glue.h:
```diff
@@ -237,39 +272,59 @@ void nrfx_busy_wait(uint32_t usec_to_wait);
 #else
 #define NRFX_PPI_CHANNELS_USED_BY_802154_DRV   0
 #define NRFX_PPI_GROUPS_USED_BY_802154_DRV     0
 #endif
```
```suggestion
#endif // defined(CONFIG_NRF_802154_RADIO_DRIVER)
```

nordic-krch commented:
@anangl comments applied.

nordic-krch commented:
@kruithofa @cvinayak can you take a look?

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5_dppi.h:
```diff
@@ -4,9 +4,12 @@
  *
  * SPDX-License-Identifier: Apache-2.0
  */
+#ifndef NORDIC_HAL_NRF5_RADIO_RADIO_NRF5_DPPI_H
+#define NORDIC_HAL_NRF5_RADIO_RADIO_NRF5_DPPI_H
```
Do not use guards in the internal controller header files.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5_dppi.h:
```diff
@@ -4,9 +4,12 @@
 
 #if defined(CONFIG_SOC_NRF5340_CPUNET) || defined(DPPI_PRESENT)
 
+#include "radio_nrf5_dppi_resources.h"
```
Move this after the list of standard includes, to after line 20.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5_dppi.h:
```diff
@@ -631,70 +598,4 @@ static inline void hal_radio_group_task_disable_ppi_setup(void)
-#else
-#define HAL_USED_PPI_GROUPS 0
-#endif
+#endif /* NORDIC_HAL_NRF5_RADIO_RADIO_NRF5_DPPI_H */
```
Remove

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5_dppi_resources.h:
```diff
@@ -0,0 +1,138 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef NORDIC_HAL_NRF5_RADIO_RADIO_NRF5_DPPI_RESOURCES_H
+#define NORDIC_HAL_NRF5_RADIO_RADIO_NRF5_DPPI_RESOURCES_H
```
Remove

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5_dppi_resources.h:
```diff
@@ -0,0 +1,138 @@
+
+#endif /* CONFIG_SOC_NRF5340_CPUNET || DPPI_PRESENT */
+
+#endif /* NORDIC_HAL_NRF5_RADIO_RADIO_NRF5_DPPI_RESOURCES_H */
```
Remove

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5_fem.h:
```diff
@@ -15,6 +15,9 @@
 #include <devicetree.h>
 #include <dt-bindings/gpio/gpio.h>
 
+#ifndef NORDIC_HAL_NRF5_RADIO_RADIO_NRF5_FEM_H__
+#define NORDIC_HAL_NRF5_RADIO_RADIO_NRF5_FEM_H__
+
```
Remove

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5_fem.h:
```diff
@@ -58,3 +61,5 @@
 #define HAL_RADIO_GPIO_LNA_POL_INV
 #endif
 #endif	/* HAL_RADIO_GPIO_HAVE_LNA_PIN */
+
+#endif /* NORDIC_HAL_NRF5_RADIO_RADIO_NRF5_FEM_H__ */
```
Remove

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5_ppi.h:
```diff
@@ -4,9 +4,12 @@
  *
  * SPDX-License-Identifier: Apache-2.0
  */
+#ifndef RADIO_NRF5_PPI_H__
+#define RADIO_NRF5_PPI_H__
```
Remove

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5_ppi.h:
```diff
@@ -4,9 +4,12 @@
 
 #if defined(CONFIG_SOC_SERIES_NRF51X) || defined(CONFIG_SOC_COMPATIBLE_NRF52X)
 
+#include "radio_nrf5_ppi_resources.h"
```
Move this after standard include to line after 14, leave a blank line in between.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5_ppi.h:
```diff
@@ -649,70 +570,4 @@ static inline void hal_radio_sw_switch_ppi_group_setup(void)
-#else
-#define HAL_USED_PPI_GROUPS 0
-#endif
+#endif /* RADIO_NRF5_PPI_H__ */
```
Remove

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5_ppi_resources.h:
```diff
@@ -0,0 +1,169 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef RADIO_NRF5_PPI_RESOURCES_H__
+#define RADIO_NRF5_PPI_RESOURCES_H__
```
Remove

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5_ppi_resources.h:
```diff
@@ -0,0 +1,169 @@
+
+#endif /* !CONFIG_BT_CTLR_TIFS_HW */
+
+#endif /* RADIO_NRF5_PPI_RESOURCES_H__ */
```
Remove

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5_resources.h:
```diff
@@ -0,0 +1,108 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef NORDIC_HAL_NRF5_RADIO_RADIO_NRF5_RESOURCES_H
+#define NORDIC_HAL_NRF5_RADIO_RADIO_NRF5_RESOURCES_H
```
Remove

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5_resources.h:
```diff
@@ -0,0 +1,108 @@
+#define BT_CTLR_USED_PPI_GROUPS 0
+#endif
+
+#endif /* NORDIC_HAL_NRF5_RADIO_RADIO_NRF5_RESOURCES_H */
```
Remove

cvinayak reviewed (changes requested):
Just use "Bluetooth: Controller: nordic : Rework hw resource files" as the title for controller changes.

nordic-krch commented:
@cvinayak thanks for review, comments applied.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5_ppi_resources.h:
```diff
@@ -0,0 +1,164 @@
+
+#else /* defined(CONFIG_BT_CTLR_TIFS_HW) || !defined(CONFIG_BT_CTLR_SW_SWITCH_SINGLE_TIMER) */
+
+#define HAL_RADIO_ENABLE_ON_TICK_PPI 6
```
Please double verify this change compared to old code

```suggestion
#define HAL_RADIO_ENABLE_ON_TICK_PPI 2
```

nordic-krch replied:
my bad. Checked other defines but did not find other error. Fixed.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5_dppi_resources.h:
```diff
@@ -0,0 +1,133 @@
+ * Note: we do not need an additional PPI, since we have already set up
+ * a PPI to publish RADIO END event.
+ */
+#define HAL_SW_SWITCH_GROUP_TASK_ENABLE_PPI HAL_SW_SWITCH_TIMER_CLEAR_PPI
```
Remove this from old code... seems to have been missed?

nordic-krch replied:
Removed. Just learned that there is no warning when define is repeated with the same value.

cvinayak reviewed (changes requested):
Sorry to not have done a better review previously. I would like @ppryga  to do another round of review incase I may have missed any more PPI assignments that could be missed/incorrect.

cvinayak reviewed (commented):
LGTM, will wait for @ppryga to approve, I believe he may be having some conflicting changes to his work in progress and would like his thoughts on the changes/review here.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5_ppi_resources.h:
```diff
@@ -0,0 +1,164 @@
+ */
+#if defined(CONFIG_SOC_NRF52805)
+#define HAL_SW_SWITCH_GROUP_TASK_DISABLE_PPI_BASE 2
+#else /* -> !defined(CONFIG_SOC_NRF52805) */
```
nit-pick, we usually write #else this way in controller:
```suggestion
#else /* !CONFIG_SOC_NRF52805 */
```

nordic-krch replied:
done

ppryga-nordic reviewed (approved):
Generally looks ok, nit-picking comments. I didn't tests it on hardware.

nordic-krch commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5_ppi_resources.h:
```diff
@@ -0,0 +1,164 @@
+ */
+#define HAL_RADIO_END_TIME_CAPTURE_PPI 27
+
+#else /* defined(CONFIG_BT_CTLR_TIFS_HW) || !defined(CONFIG_BT_CTLR_SW_SWITCH_SINGLE_TIMER) */
```
done

nordic-krch commented:
@cvinayak, @ppryga has finished his review. Can you finish your review?

nordic-krch commented:
@carlescufi hal_nordic got nrfx2.6 merged. I removed `DNM`.
